ctapipe/reco/hillas_reconstructor.py

```diff
@@ -268,56 +322,102 @@ def initialize_hillas_planes(
                 az=telescopes_pointings[tel_id].az,
                 frame=horizon_frame,
             )
 
-            camera_frame = CameraFrame(
-                focal_length=focal_length, telescope_pointing=pointing
-            )
+            focal_length = subarray.tel[tel_id].optics.equivalent_focal_length
 
-            cog_coord = SkyCoord(x=moments.x, y=moments.y, frame=camera_frame)
-            cog_coord = cog_coord.transform_to(horizon_frame)
+            if moments.x.unit.is_equivalent(u.m):  # Image parameters are in CameraFrame
 
-            p2_coord = SkyCoord(x=p2_x, y=p2_y, frame=camera_frame)
-            p2_coord = p2_coord.transform_to(horizon_frame)
+                # we just need any point on the main shower axis a bit away from the cog
+                p2_x = moments.x + 0.1 * self._cam_radius_m[camera] * np.cos(
+                    moments.psi
+                )
+                p2_y = moments.y + 0.1 * self._cam_radius_m[camera] * np.sin(
+                    moments.psi
+                )
 
-            # re-project from sky to a "fake"-parallel-pointing telescope
-            # then recalculate the psi angle
-            if self.divergent_mode:
-                camera_frame_parallel = CameraFrame(
-                    focal_length=focal_length, telescope_pointing=array_pointing
+                camera_frame = CameraFrame(
+                    focal_length=focal_length, telescope_pointing=pointing
+                )
+
+                cog_coord = SkyCoord(x=moments.x, y=moments.y, frame=camera_frame)
+                p2_coord = SkyCoord(x=p2_x, y=p2_y, frame=camera_frame)
+
+            else:  # Image parameters are already in TelescopeFrame
+
+                # we just need any point on the main shower axis a bit away from the cog
+                p2_delta_alt = moments.y + 0.1 * self._cam_radius_deg[camera] * np.sin(
+                    moments.psi
+                )
+                p2_delta_az = moments.x + 0.1 * self._cam_radius_deg[camera] * np.cos(
+                    moments.psi
+                )
+
+                telescope_frame = TelescopeFrame(telescope_pointing=pointing)
+
+                cog_coord = SkyCoord(
+                    fov_lon=moments.x, fov_lat=moments.y, frame=telescope_frame
                 )
-                cog_sky_to_parallel = cog_coord.transform_to(camera_frame_parallel)
-                p2_sky_to_parallel = p2_coord.transform_to(camera_frame_parallel)
-                angle_psi_corr = np.arctan2(
-                    cog_sky_to_parallel.y - p2_sky_to_parallel.y,
-                    cog_sky_to_parallel.x - p2_sky_to_parallel.x,
+                p2_coord = SkyCoord(
+                    fov_lon=p2_delta_az, fov_lat=p2_delta_alt, frame=telescope_frame
                 )
-                self.corrected_angle_dict[tel_id] = angle_psi_corr
+
+            # Coordinates in the sky
+            cog_coord = cog_coord.transform_to(horizon_frame)
```

Comment:
This is only needed if the reconstructed parameters are in the CameraFrame, or?

Reply:
telescope frame is not horizon frame

it is needed independently of the input frame

Reply:
In that case you can remove the "already" from line 370